File: macos/Sources/RelicCore/SaveFile.swift
import CommonCrypto
import CryptoKit
import Foundation

public enum SaveFileError: LocalizedError, Equatable, Sendable {
    case notASaveFile
    case corruptEntry(Int)

    public var errorDescription: String? {
        switch self {
        case .notASaveFile: return "不是有效的存档文件"
        case .corruptEntry(let index): return "存档条目 \(index) 越界或魔数不符"
        }
    }
}

public struct SaveRelic: Codable, Hashable, Sendable, Identifiable {
    public let index: Int
    public let itemID: Int
    public let effects: [Int]
    public let curses: [Int]

    public var id: Int { index }

    private enum CodingKeys: String, CodingKey {
        case index
        case itemID = "itemId"
        case effects
        case curses
    }

    public init(index: Int, itemID: Int, effects: [Int], curses: [Int]) {
        self.index = index
        self.itemID = itemID
        self.effects = effects
        self.curses = curses
    }
}

public struct SaveCharacter: Codable, Sendable, Identifiable {
    public let slot: Int
    public let name: String
    public let parseError: String?
    public let relics: [SaveRelic]

    public var id: Int { slot }

    public init(slot: Int, name: String, parseError: String?, relics: [SaveRelic]) {
        self.slot = slot
        self.name = name
        self.parseError = parseError
        self.relics = relics
    }
}

public struct SaveParseResult: Codable, Sendable {
    public let fileName: String
    public let checksumOk: Bool
    public let characters: [SaveCharacter]

    public init(fileName: String, checksumOk: Bool, characters: [SaveCharacter]) {
        self.fileName = fileName
        self.checksumOk = checksumOk
        self.characters = characters
    }
}

public enum SaveFileParser {
    static let aesKey: [UInt8] = [
        0x18, 0xF6, 0x32, 0x66, 0x05, 0xBD, 0x17, 0x8A,
        0x55, 0x24, 0x52, 0x3A, 0xC0, 0xA0, 0xC6, 0x09
    ]
    private static let entryMagic: [UInt8] = [0x40, 0x00, 0x00, 0x00, 0xFF, 0xFF, 0xFF, 0xFF]
    private static let faceMagic: [UInt8] = [0x27, 0x00, 0x00, 0x46, 0x41, 0x43, 0x45]
    private static let stateSlotCount = 5120
    private static let characterSlotCount = 10

    private struct SlotError: Error {
        let message: String
    }

    public static func parse(data: Data, fileName: String) throws -> SaveParseResult {
        let bytes = [UInt8](data)
        guard bytes.count >= 64, Array(bytes[0..<4]) == Array("BND4".utf8) else {
            throw SaveFileError.notASaveFile
        }
        let entryCount = Int(readInt32(bytes, 12))
        guard entryCount > 0 else {
            throw SaveFileError.notASaveFile
        }

        // 结构校验只覆盖本功能读取的前 11 个条目头（USERDATA_0..10），
        // 与 Windows 端口径一致；条目数据层面的问题在解密阶段按槽位隔离。
        let parsed = min(entryCount, characterSlotCount + 1)
        var headers: [(size: Int, offset: Int)] = []
        for i in 0..<parsed {
            let base = 64 + 32 * i
            guard base + 32 <= bytes.count else {
                throw SaveFileError.corruptEntry(i)
            }
            let size = Int(readInt32(bytes, base + 8))
            let offset = Int(readInt32(bytes, base + 16))
            guard Array(bytes[base..<base + 8]) == entryMagic,
                  size >= 0, offset > 0, offset + size <= bytes.count else {
                throw SaveFileError.corruptEntry(i)
            }
            headers.append((size, offset))
        }

        var checksumOk = true

        func decryptEntry(_ index: Int) throws -> [UInt8] {
            let header = headers[index]
            guard header.size > 16, (header.size - 16) % 16 == 0 else {
                throw SlotError(message: "条目密文长度不是 16 的倍数")
            }
            let iv = Array(bytes[header.offset..<header.offset + 16])
            let cipher = Array(bytes[header.offset + 16..<header.offset + header.size])
            let plain = try aesDecrypt(cipher: cipher, iv: iv)
            if !verifyChecksum(plain) { checksumOk = false }
            return plain
        }

        var occupied = [Bool](repeating: true, count: characterSlotCount)
        if parsed > characterSlotCount, let publicPlain = try? decryptEntry(characterSlotCount),
           let flags = slotFlags(in: publicPlain) {
            occupied = flags
        }

        var characters: [SaveCharacter] = []
        for slot in 0..<min(characterSlotCount, parsed) where occupied[slot] {
            do {
                let plain = try decryptEntry(slot)
                characters.append(parseCharacter(slot: slot, plain: plain))
            } catch {
                let message = (error as? SlotError)?.message ?? error.localizedDescription
                characters.append(SaveCharacter(
                    slot: slot,
                    name: "槽位 \(slot + 1)",
                    parseError: "该槽位解密失败：\(message)",
                    relics: []
                ))
            }
        }

        return SaveParseResult(fileName: fileName, checksumOk: checksumOk, characters: characters)
    }

    // 单槽扫描失败时保留已解析出的遗物并记入 parseError（与 Windows 端一致），
    // 不再整槽丢弃。
    private static func parseCharacter(slot: Int, plain: [UInt8]) -> SaveCharacter {
        var offset = 0x14
        var relics: [SaveRelic] = []

        for record in 0..<stateSlotCount {
            guard offset + 8 <= plain.count else {
                return SaveCharacter(
                    slot: slot,
                    name: "槽位 \(slot + 1)",
                    parseError: "存档数据截断：第 \(record) 条物品记录越界",
                    relics: relics
                )
            }
            let gaHandle = readUInt32(plain, offset)
            let length: Int
            switch gaHandle & 0xF000_0000 {
            case 0x0000_0000: length = 8
            case 0x8000_0000: length = 88
            case 0x9000_0000: length = 16
            case 0xC000_0000: length = 80
            default: length = 8
            }
            guard offset + length <= plain.count else {
                return SaveCharacter(
                    slot: slot,
                    name: "槽位 \(slot + 1)",
                    parseError: "存档数据截断：第 \(record) 条物品记录不完整",
                    relics: relics
                )
            }
            if gaHandle & 0xF000_0000 == 0xC000_0000 {
                let itemID = Int(readUInt32(plain, offset + 4) & 0x00FF_FFFF)
                let effects = [16, 20, 24].map { normalizeEffect(readUInt32(plain, offset + $0)) }
                let curses = [56, 60, 64].map { normalizeEffect(readUInt32(plain, offset + $0)) }
                relics.append(SaveRelic(index: relics.count, itemID: itemID, effects: effects, curses: curses))
            }
            offset += length
        }

        let name = characterName(plain, at: offset + 0x94) ?? "槽位 \(slot + 1)"
        return SaveCharacter(slot: slot, name: name, parseError: nil, relics: relics)
    }

    // 名字区允许截断：只读到边界内完整的 UTF-16 单元（与 Windows 端一致）。
    private static func characterName(_ plain: [UInt8], at offset: Int) -> String? {
        guard offset >= 0, offset + 2 <= plain.count else { return nil }
        var units: [UInt16] = []
        for i in 0..<16 {
            let p = offset + i * 2
            guard p + 2 <= plain.count else { break }
            let unit = UInt16(plain[p]) | (UInt16(plain[p + 1]) << 8)
            if unit == 0 { break }
            units.append(unit)
        }
        guard !units.isEmpty else { return nil }
        let name = String(decoding: units, as: UTF16.self)
        return name.isEmpty ? nil : name
    }

    // 与参考实现一致：遍历全部 FACE 魔数命中，跳过含非 0/1 字节的假阳性，
    // 用 == 1 判定占用（与 Windows 端 slotFlags 等价）。
    private static func slotFlags(in plain: [UInt8]) -> [Bool]? {
        guard plain.count >= faceMagic.count else { return nil }
        var start = 0
        while start + faceMagic.count <= plain.count {
            guard Array(plain[start..<start + faceMagic.count]) == faceMagic else {
                start += 1
                continue
            }
            let flagsStart = start - 61
            if flagsStart >= 0, flagsStart + characterSlotCount <= plain.count {
                let candidate = Array(plain[flagsStart..<flagsStart + characterSlotCount])
                if candidate.allSatisfy({ $0 <= 1 }) {
                    return candidate.map { $0 == 1 }
                }
            }
            start += 1
        }
        return nil
    }

    private static func aesDecrypt(cipher: [UInt8], iv: [UInt8]) throws -> [UInt8] {
        var plain = [UInt8](repeating: 0, count: cipher.count)
        var moved = 0
        let status = CCCrypt(
            CCOperation(kCCDecrypt),
            CCAlgorithm(kCCAlgorithmAES128),
            CCOptions(0),
            aesKey, aesKey.count,
            iv,
            cipher, cipher.count,
            &plain, plain.count,
            &moved
        )
        guard status == kCCSuccess, moved == cipher.count else {
            throw SlotError(message: "条目解密失败")
        }
        return plain
    }

    private static func verifyChecksum(_ plain: [UInt8]) -> Bool {
        guard plain.count >= 32 else { return false }
        let digest = Insecure.MD5.hash(data: Data(plain[4..<plain.count - 28]))
        return Array(digest) == Array(plain[plain.count - 28..<plain.count - 12])
    }

    private static func normalizeEffect(_ raw: UInt32) -> Int {
        (raw == 0xFFFF_FFFF || raw == 0) ? -1 : Int(raw)
    }

    private static func readUInt32(_ bytes: [UInt8], _ offset: Int) -> UInt32 {
        UInt32(bytes[offset])
            | (UInt32(bytes[offset + 1]) << 8)
            | (UInt32(bytes[offset + 2]) << 16)
            | (UInt32(bytes[offset + 3]) << 24)
    }

    private static func readInt32(_ bytes: [UInt8], _ offset: Int) -> Int32 {
        Int32(bitPattern: readUInt32(bytes, offset))
    }
}

File: windows/internal/savefile/savefile.go
// Package savefile implements read-only parsing of《黑夜君临》save archives
// (.sl2/.co2, BND4 containers with AES-128-CBC encrypted entries).
//
// The package is deliberately free of platform-specific imports so its tests
// run on any OS; the Windows shell (zenity dialog, KnownFolderPath) stays in
// package main. Nothing here ever writes to a save file.
package savefile

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	"encoding/binary"
	"errors"
	"fmt"
	"unicode/utf16"
)

// Payload is the JSON shape returned to the renderer. Field order and json
// tags are part of the bridge contract shared with the macOS implementation.
type Payload struct {
	FileName   string      `json:"fileName"`
	ChecksumOk bool        `json:"checksumOk"`
	Characters []Character `json:"characters"`
}

// Character is one occupied character slot (USERDATA_0..9).
type Character struct {
	Slot       int     `json:"slot"`
	Name       string  `json:"name"`
	ParseError *string `json:"parseError"`
	Relics     []Relic `json:"relics"`
}

// Relic is a single relic item state. Effects and Curses always hold three
// entries; empty affix slots are normalized to -1 (raw 0xFFFFFFFF or 0).
type Relic struct {
	Index   int     `json:"index"`
	ItemID  int     `json:"itemId"`
	Effects []int64 `json:"effects"`
	Curses  []int64 `json:"curses"`
}

const (
	bnd4HeaderLen      = 64
	bnd4EntryHeaderLen = 32
	ivSize             = 16

	// Per-entry integrity: MD5(plain[4 : L-28]) is stored at plain[L-28 : L-12],
	// followed by 12 bytes of padding.
	checksumStart   = 4
	checksumTrailer = 28

	characterSlots = 10   // USERDATA_0..9 are character slots; USERDATA_10 is shared.
	stateSlotCount = 5120 // fixed number of variable-length item state records
	stateStart     = 0x14 // item state records begin at this plaintext offset
	nameGap        = 0x94 // gap between the end of the state area and the name
	nameMaxUnits   = 16   // player name: UTF-16LE, at most 16 code units, NUL-terminated

	// Item entry area: 3065 fixed 14-byte records, preceded by a u32 count,
	// located 0x5B8 past the player name. Only states referenced here are in
	// the character's actual inventory — stale (deleted) relic states remain
	// in the state area and must be filtered out.
	entrySlotCount = 3065
	entryRecordLen = 14
	entryCountGap  = 0x5B8
)

var (
	bnd4Magic  = []byte("BND4")
	entryMagic = []byte{0x40, 0x00, 0x00, 0x00, 0xFF, 0xFF, 0xFF, 0xFF}
	// faceMagic locates the character-slot occupancy flags inside USERDATA_10:
	// the 10 flag bytes sit 61 bytes before a match.
	faceMagic = []byte{0x27, 0x00, 0x00, 0x46, 0x41, 0x43, 0x45}
	// aesKey is the hardcoded AES-128 key shared by every entry ("DS2 key").
	aesKey = []byte{
		0x18, 0xF6, 0x32, 0x66, 0x05, 0xBD, 0x17, 0x8A,
		0x55, 0x24, 0x52, 0x3A, 0xC0, 0xA0, 0xC6, 0x09,
	}
)

// errNotSaveFile is returned when the input is not a BND4 archive at all.
var errNotSaveFile = errors.New("不是有效的存档文件")

// Parse decodes a .sl2/.co2 archive and extracts every occupied character's
// relics. Structural corruption (bad magic, out-of-bounds entries, cipher
// length) fails the whole parse; a failure inside a single character slot is
// reported through that character's ParseError so other slots still load.
func Parse(data []byte, fileName string) (*Payload, error) {
	if len(data) < len(bnd4Magic) || !bytes.Equal(data[:len(bnd4Magic)], bnd4Magic) {
		return nil, errNotSaveFile
	}
	if len(data) < bnd4HeaderLen {
		return nil, errors.New("存档文件损坏：文件头不完整")
	}
	entryCount := int(int32(binary.LittleEndian.Uint32(data[12:16])))
	if entryCount <= 0 {
		return nil, fmt.Errorf("存档文件损坏：条目数无效（%d）", entryCount)
	}

	// Only the entries this feature reads are decoded: USERDATA_0..9 hold the
	// character slots and USERDATA_10 the shared account data.
	parsed := entryCount
	if parsed > characterSlots+1 {
		parsed = characterSlots + 1
	}

	// Structural pass: the entry header table must be intact for every entry
	// this feature reads; anything past the header table is per-entry state.
	refs := make([]entryRef, parsed)
	for i := 0; i < parsed; i++ {
		ref, err := readEntryHeader(data, i)
		if err != nil {
			return nil, err
		}
		refs[i] = ref
	}

	// Entry-level failures from here on are isolated: a broken shared entry
	// falls back to "all slots occupied", a broken character slot is reported
	// through that character's ParseError. checksumOk covers the entries this
	// parse actually consumed.
	checksumOk := true
	decrypt := func(i int) ([]byte, error) {
		plain, err := decryptEntry(data, refs[i])
		if err != nil {
			return nil, err
		}
		if !verifyChecksum(plain) {
			checksumOk = false
		}
		return plain, nil
	}

	occupied := allOccupied()
	if parsed > characterSlots {
		if shared, err := decrypt(characterSlots); err == nil {
			occupied = slotFlags(shared)
		}
	}

	characters := make([]Character, 0, characterSlots)
	for slot := 0; slot < characterSlots && slot < parsed; slot++ {
		if !occupied[slot] {
			continue
		}
		plain, err := decrypt(slot)
		if err != nil {
			msg := fmt.Sprintf("该槽位解密失败：%v", err)
			characters = append(characters, Character{
				Slot: slot, Name: fallbackName(slot), ParseError: &msg, Relics: []Relic{},
			})
			continue
		}
		characters = append(characters, parseCharacter(slot, plain))
	}

	return &Payload{FileName: fileName, ChecksumOk: checksumOk, Characters: characters}, nil
}

type entryRef struct {
	size       int
	dataOffset int
}

// readEntryHeader validates entry i's header row (magic and bounds). The
// bounds check is written overflow-safe so a crafted size/offset pair cannot
// wrap on 32-bit builds and bypass the guard.
func readEntryHeader(data []byte, i int) (entryRef, error) {
	pos := bnd4HeaderLen + bnd4EntryHeaderLen*i
	if pos+bnd4EntryHeaderLen > len(data) {
		return entryRef{}, fmt.Errorf("存档文件损坏：条目 %d 头部越界", i)
	}
	if !bytes.Equal(data[pos:pos+len(entryMagic)], entryMagic) {
		return entryRef{}, fmt.Errorf("存档文件损坏：条目 %d 魔数不符", i)
	}
	size := int(int32(binary.LittleEndian.Uint32(data[pos+8:])))
	dataOffset := int(int32(binary.LittleEndian.Uint32(data[pos+16:])))
	if size < 0 || dataOffset <= 0 || dataOffset > len(data)-size {
		return entryRef{}, fmt.Errorf("存档文件损坏：条目 %d 数据越界", i)
	}
	return entryRef{size: size, dataOffset: dataOffset}, nil
}

// decryptEntry decrypts one entry's data (IV-prefixed AES-128-CBC, no
// padding). Errors here are entry-local and isolated by the caller.
func decryptEntry(data []byte, ref entryRef) ([]byte, error) {
	if ref.size <= ivSize || (ref.size-ivSize)%aes.BlockSize != 0 {
		return nil, errors.New("加密数据长度无效")
	}
	iv := data[ref.dataOffset : ref.dataOffset+ivSize]
	ciphertext := data[ref.dataOffset+ivSize : ref.dataOffset+ref.size]
	block, err := aes.NewCipher(aesKey)
	if err != nil {
		return nil, err
	}
	plain := make([]byte, len(ciphertext))
	cipher.NewCBCDecrypter(block, iv).CryptBlocks(plain, ciphertext)
	return plain, nil
}

// verifyChecksum reports whether MD5(plain[4 : L-28]) matches the digest
// stored at plain[L-28 : L-12].
func verifyChecksum(plain []byte) bool {
	if len(plain) < checksumStart+checksumTrailer {
		return false
	}
	end := len(plain) - checksumTrailer
	sum := md5.Sum(plain[checksumStart:end])
	return bytes.Equal(sum[:], plain[end:end+md5.Size])
}

// allOccupied is the fallback occupancy table used when USERDATA_10 is
// missing or unreadable.
func allOccupied() [characterSlots]bool {
	var flags [characterSlots]bool
	for i := range flags {
		flags[i] = true
	}
	return flags
}

// slotFlags locates the 10 character-slot occupancy flags in USERDATA_10
// (61 bytes before the FACE magic). When the magic is missing, every slot
// is assumed occupied.
func slotFlags(shared []byte) [characterSlots]bool {
	flags := allOccupied()
	for pos := 0; pos+len(faceMagic) <= len(shared); {
		hit := bytes.Index(shared[pos:], faceMagic)
		if hit < 0 {
			break
		}
		start := pos + hit - 61
		if start >= 0 && start+characterSlots <= len(shared) && flagBytesValid(shared[start:start+characterSlots]) {
			for i := 0; i < characterSlots; i++ {
				flags[i] = shared[start+i] == 1
			}
			return flags
		}
		pos += hit + 1
	}
	return flags
}

// flagBytesValid rejects candidate flag runs containing anything but 0/1,
// mirroring the reference implementation's false-positive guard.
func flagBytesValid(b []byte) bool {
	for _, v := range b {
		if v > 1 {
			return false
		}
	}
	return true
}

// parseCharacter scans one decrypted character slot: 5120 variable-length
// item state records starting at 0x14, then the player name 0x94 past the
// end of the state area. Truncation is reported via ParseError, never panic.
func parseCharacter(slot int, plain []byte) Character {
	c := Character{Slot: slot, Name: fallbackName(slot), Relics: []Relic{}}
	fail := func(format string, args ...interface{}) Character {
		msg := fmt.Sprintf(format, args...)
		c.ParseError = &msg
		return c
	}

	off := stateStart
	gaHandles := []uint32{}
	for rec := 0; rec < stateSlotCount; rec++ {
		if off+8 > len(plain) {
			return fail("存档数据截断：第 %d 条物品记录越界", rec)
		}
		gaHandle := binary.LittleEndian.Uint32(plain[off:])
		itemID := binary.LittleEndian.Uint32(plain[off+4:])

		// Record length by type nibble; unknown non-empty types (e.g. goods,
		// 0xB...) occupy the 8 header bytes only.
		size := 8
		switch gaHandle & 0xF0000000 {
		case 0x80000000: // weapon
			size = 88
		case 0x90000000: // armor
			size = 16
		case 0xC0000000: // relic
			size = 80
		}
		if off+size > len(plain) {
			return fail("存档数据截断：第 %d 条物品记录不完整", rec)
		}
		if gaHandle&0xF0000000 == 0xC0000000 {
			gaHandles = append(gaHandles, gaHandle)
			c.Relics = append(c.Relics, Relic{
				Index:  len(c.Relics),
				ItemID: int(itemID & 0x00FFFFFF),
				Effects: []int64{
					affixAt(plain, off+16),
					affixAt(plain, off+20),
					affixAt(plain, off+24),
				},
				Curses: []int64{
					affixAt(plain, off+56),
					affixAt(plain, off+60),
					affixAt(plain, off+64),
				},
			})
		}
		off += size
	}

	nameOff := off + nameGap
	if name, ok := readName(plain, nameOff); ok {
		c.Name = name
	}
	c.Relics = filterOwnedRelics(plain, nameOff, c.Relics, gaHandles)
	return c
}

// filterOwnedRelics keeps only relics whose gaHandle is referenced by the
// item entry area — deleted relics leave stale records in the state area.
// When the entry area is unreadable the list is returned unfiltered.
func filterOwnedRelics(plain []byte, nameOff int, relics []Relic, gaHandles []uint32) []Relic {
	countOff := nameOff + entryCountGap
	if countOff < 0 || countOff+4 > len(plain) {
		return relics
	}
	owned := make(map[uint32]bool)
	readable := false
	for slot := 0; slot < entrySlotCount; slot++ {
		pos := countOff + 4 + slot*entryRecordLen
		if pos+entryRecordLen > len(plain) {
			break
		}
		readable = true
		ga := binary.LittleEndian.Uint32(plain[pos:])
		if ga&0xF0000000 == 0xC0000000 {
			owned[ga] = true
		}
	}
	if !readable {
		return relics
	}
	kept := make([]Relic, 0, len(relics))
	for i, relic := range relics {
		if owned[gaHandles[i]] {
			relic.Index = len(kept)
			kept = append(kept, relic)
		}
	}
	return kept
}

// fallbackName is the display name used when a slot's real name cannot be
// read; 1-based to match how the UI numbers slots.
func fallbackName(slot int) string {
	return fmt.Sprintf("槽位 %d", slot+1)
}

// affixAt reads an affix ID and normalizes both empty sentinels
// (0xFFFFFFFF and 0) to -1.
func affixAt(plain []byte, off int) int64 {
	v := binary.LittleEndian.Uint32(plain[off:])
	if v == 0xFFFFFFFF || v == 0 {
		return -1
	}
	return int64(v)
}

// readName decodes the UTF-16LE player name (at most 16 code units,
// NUL-terminated). It reports false when the region is unreadable or empty,
// letting the caller keep the "槽位 N" fallback.
func readName(plain []byte, off int) (string, bool) {
	if off < 0 || off+2 > len(plain) {
		return "", false
	}
	units := make([]uint16, 0, nameMaxUnits)
	for i := 0; i < nameMaxUnits; i++ {
		p := off + 2*i
		if p+2 > len(plain) {
			break
		}
		u := binary.LittleEndian.Uint16(plain[p:])
		if u == 0 {
			break
		}
		units = append(units, u)
	}
	if len(units) == 0 {
		return "", false
	}
	return string(utf16.Decode(units)), true
}
